src/Builder/Encoder/RecursiveEncode.php

```diff
@@ -4,15 +4,18 @@
 
 namespace MongoDB\Builder\Encoder;
 
-use MongoDB\Builder\BuilderEncoder;
+use MongoDB\Codec\Encoder;
 use stdClass;
+use WeakReference;
 
 use function get_object_vars;
 use function is_array;
 
+/** @internal */
 trait RecursiveEncode
 {
-    final public function __construct(protected readonly BuilderEncoder $encoder)
+    /** @param WeakReference<Encoder> $encoder */
+    final public function __construct(private readonly WeakReference $encoder)
```

Comment:
Is there a way to enforce the type of the class the `WeakReference` is pointing to? Would that entail checking the type of `$encoder->get()` and manually throwing a `TypeError` if it's wrong?

I don't think we need to do this here since this is an internal trait, mainly just asking out of curiosity.

Reply:
You would like native support for generics, that doesn't exist.
Also, the `WeakReference` class is final, we cannot extend it to restrict the type.